File: apps/ios/RafayPair/Core/Pose/PoseGeometry.swift
import Foundation

/// Geometry primitives for the pose engine.
///
/// The expression order matches `packages/pose-engine/src/geometry.ts`
/// literally. Addition, subtraction, multiplication, division, and square root
/// are exactly reproducible across platforms; `atan2` is not guaranteed to be,
/// which is why golden vectors compare continuous values with a tolerance.
enum PoseGeometry {
    struct Point: Sendable, Equatable {
        var x: Double
        var y: Double
    }

    private static let degreesPerRadian = 180.0 / Double.pi
    private static let minVectorMagnitude = 1e-9

    static func midpoint(_ a: Point, _ b: Point) -> Point {
        Point(x: (a.x + b.x) / 2, y: (a.y + b.y) / 2)
    }

    static func distance(_ a: Point, _ b: Point) -> Double {
        let dx = a.x - b.x
        let dy = a.y - b.y
        return (dx * dx + dy * dy).squareRoot()
    }

    /// Unsigned angle in degrees between vectors, in `0...180`.
    ///
    /// `atan2(|cross|, dot)` is used rather than the cosine form because it
    /// stays accurate for nearly parallel and nearly antiparallel vectors —
    /// exactly the extended-limb angles the classifier depends on.
    static func angleBetween(_ a: Point, _ b: Point) -> Double {
        let magnitudeA = (a.x * a.x + a.y * a.y).squareRoot()
        let magnitudeB = (b.x * b.x + b.y * b.y).squareRoot()
        guard magnitudeA >= minVectorMagnitude, magnitudeB >= minVectorMagnitude else {
            return 180
        }
        let cross = abs(a.x * b.y - a.y * b.x)
        let dot = a.x * b.x + a.y * b.y
        return atan2(cross, dot) * degreesPerRadian
    }

    /// Angle at `vertex` subtended by `a` and `b`, in `0...180`.
    static func angleAtVertex(_ a: Point, _ vertex: Point, _ b: Point) -> Double {
        angleBetween(
            Point(x: a.x - vertex.x, y: a.y - vertex.y),
            Point(x: b.x - vertex.x, y: b.y - vertex.y)
        )
    }
}

extension Joint {
    var point: PoseGeometry.Point {
        PoseGeometry.Point(x: x, y: y)
    }
}

File: apps/android/app/src/main/java/com/rafaypair/android/pose/PoseGeometry.kt
package com.rafaypair.android.pose

import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.sqrt

/**
 * Geometry primitives for the pose engine.
 *
 * The expression order matches `packages/pose-engine/src/geometry.ts` literally.
 * Addition, subtraction, multiplication, division, and square root are exactly
 * reproducible across platforms; `atan2` is not guaranteed to be, which is why
 * golden vectors compare continuous values with a tolerance.
 */
object PoseGeometry {
    data class Point(val x: Double, val y: Double)

    private const val DEGREES_PER_RADIAN = 180.0 / Math.PI
    private const val MIN_VECTOR_MAGNITUDE = 1e-9

    fun midpoint(a: Point, b: Point): Point = Point((a.x + b.x) / 2, (a.y + b.y) / 2)

    fun distance(a: Point, b: Point): Double {
        val dx = a.x - b.x
        val dy = a.y - b.y
        return sqrt(dx * dx + dy * dy)
    }

    /**
     * Unsigned angle in degrees between vectors, in `0..180`.
     *
     * `atan2(|cross|, dot)` is used rather than the cosine form because it stays
     * accurate for nearly parallel and nearly antiparallel vectors — exactly the
     * extended-limb angles the classifier depends on.
     */
    fun angleBetween(a: Point, b: Point): Double {
        val magnitudeA = sqrt(a.x * a.x + a.y * a.y)
        val magnitudeB = sqrt(b.x * b.x + b.y * b.y)
        if (magnitudeA < MIN_VECTOR_MAGNITUDE || magnitudeB < MIN_VECTOR_MAGNITUDE) {
            return 180.0
        }
        val cross = abs(a.x * b.y - a.y * b.x)
        val dot = a.x * b.x + a.y * b.y
        return atan2(cross, dot) * DEGREES_PER_RADIAN
    }

    /** Angle at [vertex] subtended by [a] and [b], in `0..180`. */
    fun angleAtVertex(a: Point, vertex: Point, b: Point): Double = angleBetween(
        Point(a.x - vertex.x, a.y - vertex.y),
        Point(b.x - vertex.x, b.y - vertex.y),
    )
}

internal fun Joint.point(): PoseGeometry.Point = PoseGeometry.Point(x, y)
